### PR TITLE
Send Remind/FollowUp emails in choosen time [#299]

### PR DESCRIPTION
**Pull request by hekeemje**

### CODE REVIEW

**Review by va-dym-ko-val — APPROVED** [review]

**Comment by hekeemje** [comment]
image

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayJob.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            _provider = provider;
+        }
+
+        public Task Execute(IJobExecutionContext context)
+        {
+            using var scope = _provider.CreateScope();
+            var emailDelayService = scope.ServiceProvider.GetRequiredService<IEmailDelayService>();
+
+            Task.Run(() => emailDelayService.CheckForNotify()).Wait();
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayJob.cs]: Task.Run won't give you async execusion

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        private readonly IMeetingService _meetingService; 
+
+        public EmailDelayService(EasyMeetsCoreContext context, IMapper mapper, IMeetingService meetingService) : base(context, mapper)
+        {
+            _meetingService = meetingService;
+        }
+
+        public async Task CheckForNotify()
+        {
+            var avslots = _context.AvailabilitySlots.ToList();
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs]: not good idea to load all slots

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                foreach (var meeting in slot.Meetings)
+                {
+                    double minutes = TimeSpan.Parse(emailTemplate.TimeValue).TotalMinutes * -1;
+
+                    if (meeting.StartTime.AddMinutes(minutes) == DateTimeOffset.Now)
+                    {
+                        await _meetingService.SendEmailsAsync(meeting.Id, TemplateType.Reminders);
+                    }
+                }
+            }
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs]: foreach in foreach is not a good idea

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                }
+
+                foreach (var meeting in slot.Meetings)
+                {
+                    double minutes = TimeSpan.Parse(emailTemplate.TimeValue).TotalMinutes;
+
+                    if (meeting.StartTime.AddMinutes(minutes) == DateTimeOffset.Now)
+                    {
+                        await _meetingService.SendEmailsAsync(meeting.Id, TemplateType.FollowUp);
+                    }
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs]: see the previous comment

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+        public EmailDelayService(EasyMeetsCoreContext context, IMapper mapper, IMeetingService meetingService) : base(context, mapper)
+        {
+            _meetingService = meetingService;
+        }
+
+        public async Task CheckForNotify()
+        {
+            var slots = await _context.EmailTemplates
+                .Where(x => (x.TemplateType == TemplateType.Reminders && x.IsSend == true) || (x.TemplateType == TemplateType.FollowUp && x.IsSend == true))
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs]:
```suggestion
                .Where(x => (x.TemplateType == TemplateType.Reminders && x.IsSend) || (x.TemplateType == TemplateType.FollowUp && x.IsSend))
```

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                .Include(x => x.AvailabilitySlot)
+                .Select(x => x.AvailabilitySlot).ToListAsync();
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs]:
```suggestion
                .Select(x => x.AvailabilitySlot)
                .ToListAsync();
```

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+            await Notify(slots, TemplateType.Reminders);
+            await Notify(slots, TemplateType.FollowUp);
+
+            await Task.CompletedTask;
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs]: You can use task.whenall

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                }
+
+                double minutes = TimeSpan.Parse(emailTemplate.TimeValue).TotalMinutes;
+
+                if(templateType == TemplateType.Reminders)
+                {
+                    minutes *= -1;
+                }
+
+                _ = slot.Meetings.Select(async x => await CheckAndSend(minutes, x, templateType));
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/Quartz/EmailDelayService.cs]: It would be better to wait for all not one simple template

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Review by apofis96 — APPROVED** [review]

**Pull request merged**